bitcoin/src/blockdata/transaction.rs

```diff
@@ -530,6 +550,12 @@ impl TxOut {
     }
 }
 
+/// Returns the total number of bytes that this script pubkey would contribute to a transaction.
+fn size_from_script_pubkey(script_pubkey: &Script) -> usize {
+    let len = script_pubkey.len();
+    Amount::SIZE + VarInt::from(len).size() + len
+}
```

Comment:
In 3d69a9bf832663bf50dc25a748bed9407a774ad3:

Same story: this will panic if the script size is within a few bytes of 2^64. We should just document it and move on.

Reply:
Is it worth introducing an invariant on the `ScriptBuf` and `Script` types to limit the size?

A quick websearch brought up this https://bitcoin.stackexchange.com/questions/117594/what-are-bitcoins-transaction-and-script-limits#117595

In case anyone is not using it this search site is awesome: https://bitcoinsearch.xyz/

Reply:
Yeah, I think we should restrict `Script` to have a maximum 2^31 size.

Reply:
Leaving this for another day because I don't want to hackishly add an invariant and I don't feel like thinking hard about it right now. I used saturating add/mul to resolve this issue.

Reply:
Same comment as I made elsewhere, but just saturating can be bad IMO since it leads to a silent failure.  I agree with @apoelstra here to just panic and mark as panic and move on if wrapping in an error or none is too heavy to do now.

Reply:
@apoelstra why limit the script? I don't remember this limit in consensus rules. Also I suspect it could make using scripts much more annoying.

Reply:
@Kixunil in practice, scripts have to fit into blocks, which are limited to 4M.

Currently we have limited script to "whatever the maximum number of elements a `Vec` can hold" which is at most 2^32 on 32-bit systems, but which is likely to be 2^31 in practice because 32-bit allocators want to be able to hold allocation-sized signed diffs. These system-specific limits require us to have overflow checks all over the place when manipulating scripts, such as in this case where we are trying to add 12 to a script size and are worried that we'll overflow `usize`.

If we had a 2^31 limit this would:
* Never be encountered in practice.
* Not affect the API for any of the "template constructors" like p2wsh, or conversions from addresses, or parsing from transactions (which I believe already have a "maximum vector size" check which has a much lower limit).
* In the case of `script::Builder` we could defer the API impact to `into_script` so that users could continue to use `push_*` whatever without needing to check errors constantly, and only check for overflow in the end. (And we could provide a panicking variant for users who were sure they weren't going to overflow.)

Reply:
We would also have to limit input and output counts and transaction counts, since you could technically have so many of them it overflows. I'm not convinced putting all these checks everywhere is better than overflow checks everywhere. Overflow checks are at least pretty easy to understand and get right.

Reply:
This is true -- trying to bound all types to stay with in a non-overflowing window would be really hard (and probably impossible in some cases; e.g. you can have a taptree with 2^128 leaves in theory which would still be legal to use on-chain).